compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -562,9 +562,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     self.suggest_semicolon_at_end(expr.span, err);
                 });
             }
-            hir::StmtKind::Semi(ref expr) => {
-                self.check_expr(&expr);
-            }
+            hir::StmtKind::Semi(ref expr) => match expr.kind {
+                hir::ExprKind::Match(discrim, arms, source) if is_last => {
+                    // All of this is equivalent to calling `check_expr`, but it is inlined out here
+                    // in order to capture the fact that this `match` is the last statement in its
+                    // function. This is done for better suggestions to remove the `;`.
```

Comment:
Maybe we should instead have a new `Expectation` variant: `Expectation::IsLast`, then you can call `check_expr_with_expectation`, which should hopefully cut down on any duplication

Reply:
Done.